frontend/src/Components/ContentComponent.tsx

```diff
@@ -56,6 +78,123 @@ function updateContent(div: HTMLDivElement, setZoomedImg: (img: ZoomedImg | null
     div.querySelectorAll('details.expand').forEach(expand => {
         updateExpand(expand as HTMLDetailsElement);
     });
+
+    div.querySelectorAll('span.secret-mailbox').forEach(mailbox => {
+        updateMailbox(mailbox as HTMLSpanElement, setMailboxKey);
+    });
+
+    div.querySelectorAll('span.secret-mail').forEach(mail => {
+        updateMail(mail as HTMLSpanElement, setMailboxKey, currentUsername);
+    });
+}
+
+function updateMailbox(mailbox: HTMLSpanElement,
+                          setMailboxKey: (key: MailboxKey | null) => void
+                       ) {
+    const secret = mailbox.dataset.secret;
+    if (!secret) {
+        return;
+    }
+    let mailboxTitle = mailbox.dataset.rawText;
+    try {
+        mailboxTitle = mailboxTitle && b64DecodeUnicode(mailboxTitle);
+    } catch (e) {
+        mailboxTitle = undefined;
+    }
+
+    mailbox.addEventListener('click', () => {
+        setMailboxKey({
+            type: 'mailbox',
+            openKey: secret,
+            mailboxTitle
+        });
+    });
+}
+
+function updateMail(mail: HTMLSpanElement,
+                    setMailboxKey: (key: MailKey | null) => void,
+                    currentUsername?: string
+) {
+    // check processed
+    if (mail.dataset.processed) {
+        return;
+    }
+    mail.dataset.processed = '1';
+
+    const secret = mail.dataset.secret;
+    if (!secret) {
+        mail.classList.add('secret-mail-error');
+        return;
+    }
+    let cipher: string | undefined;
+    let encodedKey: string;
+
+    // try decode secret as json
+    try {
+        const j = JSON.parse(b64DecodeUnicode(secret));
+        // add mention "для @username"
+        if (j.to && !mail.querySelector('span.mention')) {
+            const mention = document.createElement('span');
+            mention.classList.add('mention');
+            mention.innerText = j.to;
+            mail.appendChild(document.createTextNode(' для '));
+            mail.appendChild(mention);
+        }
+
+        // check v, to, c
+        if (!j.v || !j.c || !Number.isInteger(j.v) || !j.toKey) {
+            throw new Error('Invalid secret');
+        } else if (j.to && j.to === currentUsername && j.toKey) {
+            encodedKey = j.toKey;
+            cipher = j.c;
+        } else if (j.from && j.from === currentUsername && j.fromKey) {
+            encodedKey = j.fromKey;
+            cipher = j.c;
+        } else if (!j.to) {
+            encodedKey = j.toKey;
+            cipher = j.c;
+        }
+
+        if (!cipher) {
+            mail.classList.add('secret-mail-disabled');
+            return;
+        }
+    } catch (e) {
+        // add error class
+        mail.classList.add('secret-mail-error');
+        return;
+    }
+
+    // just the text
+    const title = mail.innerText.trim();
+
+    const mailInnerHtml = mail.innerHTML;
+    let decoded = false;
+
+    mail.addEventListener('click', () => {
+        if (decoded || !cipher) {
```

Comment:
I think this check can be done above with early return so that event handler would be just not set at all

Reply:
Can add an extra check, but variables (`let`) have to be double-checked in the closure.

Reply:
Nevermind that. If you look carefully, `cipher` is already checked before, and `decoded` can't be true before the callback.